Cycle after resolution of compiler upgrade (2.1.0~beta2)

I believe this is a bug in the function that alters the action graph in orde to push uninstall to the latest possible moment (`OpamCudf.ActionGraph.explicit`).; logging the case here so we can reproduce later.

repo revision: 3c21183b

```
$ opam sw cr /tmp/testsw ocaml-variants.4.10.0+flambda --fake
$ opam install --sw /tmp/testsw ocaml-base-compiler.4.11.1 --fake --update-invariant
The following actions will be faked:
  ⊘ remove    ocaml-variants      4.10.0+flambda   [conflicts with ocaml-base-compiler]
  ∗ install   ocaml-base-compiler 4.11.1
  ↻ recompile ocaml-config        1                [uses ocaml-base-compiler]
  ↗ upgrade   ocaml               4.10.0 to 4.11.1 [uses ocaml-base-compiler]
===== ∗ 1   ↻ 1   ↗ 1   ⊘ 1 =====
Do you want to continue? [Y/n] y

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] Cycles found during dependency resolution:
          - ⊘ ocaml-config.1 -> ⊘ ocaml-variants.4.10.0+flambda -> λ ocaml-base-compiler.4.11.1 -> ⊘
        ocaml.4.10.0
```
### Base action graph
![base-actions](https://user-images.githubusercontent.com/807966/93082471-482fbb80-f691-11ea-9560-a9082d5ba507.png)

### Expanded action graph
![explicit-actions](https://user-images.githubusercontent.com/807966/93082479-4b2aac00-f691-11ea-84e9-6238ca595bb2.png)

Presumably, the edge from `build ocaml-base-compiler.4.11.1` to `remove ocaml.4.10.0` was wrongly added by this function.

--- a/master_changes.md
+++ b/master_changes.md
@@ -78,7 +78,7 @@ New option/command/subcommand are prefixed with ◈.
   *
 
 ## Internal
-  *
+  * ActionGraph: removal postponing, protect against addition of cycles [#4358 @AltGr - fix #4357]
 
 ## Test
   *

--- a/src/solver/opamActionGraph.ml
+++ b/src/solver/opamActionGraph.ml
@@ -238,6 +238,8 @@ module Make (A: ACTION) : SIG with type package = A.package = struct
       ) !reduced;
     g
 
+  let same_name p1 p2 = A.Pkg.(name_to_string p1 = name_to_string p2)
+
   let compute_closed_predecessors noop_remove g =
     let closed_g = copy g in
     transitive_closure closed_g;
@@ -264,7 +266,12 @@ module Make (A: ACTION) : SIG with type package = A.package = struct
             let preds =
               List.filter
                 (function
-                  | `Build p -> Set.mem p closed_packages
+                  | `Build q as b ->
+                    Set.mem q closed_packages &&
+                    not (List.exists (function
+                        | `Remove r -> same_name p r
+                        | _ -> false)
+                        (pred closed_g b))
                   | _ -> false)
                 (pred closed_g a) in
             OpamStd.String.Map.add (A.Pkg.name_to_string p) preds acc
@@ -280,7 +287,6 @@ module Make (A: ACTION) : SIG with type package = A.package = struct
 
   let explicit ?(noop_remove = (fun _ -> false)) ~sources_needed g0 =
     let g = copy g0 in
-    let same_name p1 p2 = A.Pkg.(name_to_string p1 = name_to_string p2) in
     (* We insert a "build" action before any "install" action.
        Except, between the removal and installation of the same package
        (the removal might be postponed after a succesfull build. *)

